src/main/java/org/mule/service/oauth/internal/AbstractOAuthDancer.java

```diff
@@ -138,6 +164,95 @@ protected String handleClientCredentials(final Map<String, String> formData) {
+
+    // If the context was just created, initialize it.
+    if (oauthContext.getDancerState() == NO_TOKEN) {
+      if (lock.tryLock()) {
+        try {
+          oauthContext = oauthContextSupplier.get();
+          if (oauthContext.getDancerState() == HAS_TOKEN) {
+            // Some other thread/node completed the refresh before lock was acquired here. Very quickly and quite improbable, but
+            // possible.
+            return completedFuture(null);
+          } else if (oauthContext.getDancerState() == REFRESHING_TOKEN) {
+            return pollForRefreshComplete(oauthContextSupplier);
+          } else if (oauthContext.getDancerState() == NO_TOKEN) {
+            ((DefaultResourceOwnerOAuthContext) oauthContext).setDancerState(REFRESHING_TOKEN);
+            updateResourceOwnerOAuthContext(oauthContext);
+
+            return tokenRefreshRequester.apply(oauthContext);
+          }
+        } finally {
+          lock.unlock();
+        }
+      } else {
+        return pollForRefreshComplete(oauthContextSupplier);
+      }
+    }
+
+    // If there is a previous token, refresh it
+    if (oauthContext.getDancerState() == HAS_TOKEN) {
+      final String accessToken = oauthContext.getAccessToken();
+      lock.lock();
+      try {
+        oauthContext = oauthContextSupplier.get();
+        if (oauthContext.getDancerState() == HAS_TOKEN) {
+          if (accessToken.equals(oauthContext.getAccessToken())) {
+            ((DefaultResourceOwnerOAuthContext) oauthContext).setDancerState(REFRESHING_TOKEN);
+            updateResourceOwnerOAuthContext(oauthContext);
+
+            return tokenRefreshRequester.apply(oauthContext);
+          } else {
+            // Some other thread/node completed the refresh before lock was acquired here. Very quickly and quite improbable, but
+            // possible.
+            return completedFuture(null);
+          }
+        } else if (oauthContext.getDancerState() == REFRESHING_TOKEN) {
+          return pollForRefreshComplete(oauthContextSupplier);
+        } else if (oauthContext.getDancerState() == NO_TOKEN) {
+          ((DefaultResourceOwnerOAuthContext) oauthContext).setDancerState(REFRESHING_TOKEN);
+          updateResourceOwnerOAuthContext(oauthContext);
+
+          return tokenRefreshRequester.apply(oauthContext);
+        }
+      } finally {
+        lock.unlock();
+      }
+    }
+
+    // In any other case, a refresh is being done elsewhere, so we poll for it
+    return pollForRefreshComplete(oauthContextSupplier);
+  }
+
+  private CompletableFuture<Void> pollForRefreshComplete(Supplier<ResourceOwnerOAuthContext> oauthContextSupplier) {
+    final CompletableFuture<Void> pendingResponse = new CompletableFuture<>();
+
+    pollScheduler.execute(() -> {
+      DefaultResourceOwnerOAuthContext ctx = (DefaultResourceOwnerOAuthContext) oauthContextSupplier.get();
+
+      while (ctx.getDancerState() == REFRESHING_TOKEN) {
+        try {
+          sleep(10);
```

Comment:
too short of a frequency. In most cases this is waiting for an HTTP response hitting a service provider in the internet. I wouldn't use a number under 100